src/storage/src/source/postgres.rs

```diff
@@ -525,184 +525,207 @@ async fn postgres_replication_loop(mut task_info: PostgresTaskInfo) {
     }
 }
 
-/// Core logic
-async fn postgres_replication_loop_inner(
+/// The logic to snapshot the PG collections.
+///
+/// Importantly this is broken out so we can ensure nothing returns an indefinite error, which can
+/// happen inadvertently.
+async fn postgres_replication_loop_inner_snapshot(
     task_info: &mut PostgresTaskInfo,
 ) -> Result<(), ReplicationError> {
-    if task_info.replication_lsn == PgLsn::from(0) {
-        // Get all the relevant tables for this publication
-        let publication_tables = mz_postgres_util::publication_info(
-            &task_info.connection_config,
-            &task_info.publication,
-            None,
-        )
+    // Get all the relevant tables for this publication
+    let publication_tables = mz_postgres_util::publication_info(
+        &task_info.connection_config,
+        &task_info.publication,
+        None,
+    )
+    .await
+    .err_indefinite()?;
+
+    // Validate publication tables against the state snapshot
+    determine_table_compatibility(&task_info.source_tables, publication_tables).err_definite()?;
+
+    let client = task_info
+        .connection_config
+        .clone()
+        .connect_replication()
         .await
         .err_indefinite()?;
 
-        // Validate publication tables against the state snapshot
-        determine_table_compatibility(&task_info.source_tables, publication_tables)
-            .err_definite()?;
+    // Technically there is TOCTOU problem here but it makes the code easier and if we end
+    // up attempting to create a slot and it already exists we will simply retry
+    // Also, we must check if the slot exists before we start a transaction because creating a
+    // slot must be the first statement in a transaction
+    let res = client
+        .simple_query(&format!(
+            r#"SELECT confirmed_flush_lsn FROM pg_replication_slots WHERE slot_name = '{}'"#,
+            task_info.slot
+        ))
+        .await?;
+    let slot_lsn = parse_single_row(&res, "confirmed_flush_lsn");
+    client
+        .simple_query("BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ;")
+        .await?;
```

Comment:
do we know what error was being erroneously marked as indefinite instead of irrecoverable? Should we get rid of this: https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/postgres.rs#L147-L155 impl and force every error in this module to explicitly use `map_err`?

Reply:
That was my original approach but that still let people mess up, where as the try-catch is much harder to misuse, and with that in place it matters less. idk; I'm equivocal.